src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,598 @@
+
+                // Can't use ArrayPool for the below array since it's an instance field of this object.
+                // But since we never expose the raw array contents to our callers we can get away
+                // with skipping the array zero-init during allocation. The segment points to the
+                // data which we haven't yet read; however, we own the entire backing array and can
+                // re-create the segment as needed once the array is repopulated.
+
+                _pendingReadData = new ArraySegment<byte>(GC.AllocateUninitializedArray<byte>(_thisEncoding.GetMaxByteCount(_readCharBufferMaxSize)), 0, 0);
+            }
+        }
+
+        // Sets up the data structures that are necessary before any write operation takes place,
+        // throwing if the object is in a state where writes are not possible.
+        [MemberNotNull(nameof(_innerStream), nameof(_thisDecoder), nameof(_innerEncoder))]
+        private void EnsurePreWriteConditions()
+        {
+            ThrowIfDisposed();
+            if (_innerEncoder is null)
+            {
+                InitializeReadDataStructures();
+            }
+
+            void InitializeReadDataStructures()
+            {
+                if (!CanWrite)
+                {
+                    throw Error.GetWriteNotSupported();
+                }
+
+                _innerEncoder = _innerEncoding.GetEncoder();
+                _thisDecoder = _thisEncoding.GetDecoder();
+            }
+        }
+
+#pragma warning restore CS8774 // Member must have a non-null value when exiting.
+#pragma warning restore CS3016 // Arrays as attribute arguments is not CLS-compliant
+
+        // returns any pending data that needs to be flushed to the inner stream before disposal
+        private ArraySegment<byte> FinalFlushWriteBuffers()
+        {
+            // If this stream was never used for writing, no-op.
+
+            if (_thisDecoder is null || _innerEncoder is null)
+            {
+                return default;
+            }
+
+            // Having leftover data in our buffers should be very rare since it should only
+            // occur if the end of the stream contains an incomplete multi-byte sequence.
+            // Let's not bother complicating this logic with array pool rentals or allocation-
+            // avoiding loops.
+
+            // convert bytes [this] -> chars
+
+            char[] chars = Array.Empty<char>();
+            int charCount = _thisDecoder.GetCharCount(Array.Empty<byte>(), 0, 0, flush: true);
+            if (charCount > 0)
+            {
+                chars = new char[charCount];
+                charCount = _thisDecoder.GetChars(Array.Empty<byte>(), 0, 0, chars, 0, flush: true);
+            }
+
+            // convert chars -> bytes [inner]
+
+            byte[] bytes = Array.Empty<byte>();
+            int byteCount = _innerEncoder.GetByteCount(chars, 0, charCount, flush: true);
```

Comment:
Any benefit to wrapping this and the subsequent if in an `if charCount > 0` check?  And/or moving all of this up into the previous `if (charCount > 0)` block?  Based on the previous comment, it sounds like we want to optimize for the case where charCount == 0.

Reply:
They need to be checked separately. It's possible that *_innerEncoder* needs to perform some end-of-data fixup, even if *_thisDecoder* doesn't. The only way to tell is to pass _flush: true_ to each API and to see if it reports back that any new data needs to be written. I'll clarify this in a code comment.